# Managing large and complex training datasets

Large and complex datasets are difficult to use for training models due to computational limitations (enormous amounts of RAM are required just to open such datasets). <br>

Additionally, the time needed to select genes for model training increases. <br>

Meanwhile, increasing the amount of data used for training does not always significantly improve model quality.

Here, we present a method to overcome computational limits for training models on large, complex datasets with many donors.

In [1]:
# Python packages
import warnings
warnings.simplefilter('ignore')

import scanpy as sc
import scparadise
import numpy as np
import pandas as pd
import os

sc.set_figure_params(dpi = 120)

In [2]:
# Create folder to save files
# single nuclear RNA-seq of human retina
os.makedirs('snRNAseq_human_retina', exist_ok=True)

In [3]:
# Download CELLxGENE dataset (snRNA-seq of human retina): 
# https://cellxgene.cziscience.com/collections/4c6eaf5c-6d57-4c76-b1e9-60df8c655f1e
!wget https://datasets.cellxgene.cziscience.com/2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad

--2025-02-08 13:49:05--  https://datasets.cellxgene.cziscience.com/2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad
Resolving datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)... 52.85.49.24, 52.85.49.28, 52.85.49.17, ...
Connecting to datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)|52.85.49.24|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 37946797973 (35G) [binary/octet-stream]
Saving to: ‘2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad’

2e910e62-7eaf-4c06- 100%[===================>]  35.34G  25.1MB/s    in 27m 20s 

2025-02-08 14:13:58 (22.1 MB/s) - ‘2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad’ saved [37946797973/37946797973]



In [4]:
# Check metadata using backed mode
adata = sc.read_h5ad('2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad', backed = 'r')
adata

AnnData object with n_obs × n_vars = 3177310 × 36406 backed at '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad'
    obs: 'reference_genome', 'gene_annotation_version', 'alignment_software', 'intronic_reads_counted', 'donor_id', 'donor_age', 'self_reported_ethnicity_ontology_term_id', 'donor_cause_of_death', 'donor_living_at_sample_collection', 'organism_ontology_term_id', 'sample_id', 'sample_preservation_method', 'tissue_ontology_term_id', 'development_stage_ontology_term_id', 'sample_collection_method', 'tissue_source', 'tissue_type', 'suspension_derivation_process', 'suspension_dissociation_reagent', 'suspension_enriched_cell_types', 'suspension_enrichment_factors', 'suspension_uuid', 'suspension_type', 'tissue_handling_interval', 'library_id', 'assay_ontology_term_id', 'sequenced_fragment', 'institute', 'library_id_repository', 'sequencing_platform', 'is_primary_data', 'cell_type_ontology_term_id', 'author_cell_type', 'disease_ontology_term_id', 'sex_ontology_term_id', 'majorclass', 'AC

## Train a scAdam model using a fraction of the dataset
The entire dataset contains 3,177,310 cells and 36,406 genes (35.34 GB), which is too large to open on a standard computer. <br>
In addition, selecting genes and training a model on a dataset of this size require substantial computational time and resources.

Therefore, the scParadise team recommends extracting a smaller subset of the dataset for scAdam model training.

In [5]:
# Obtain 25000 cells from dataset
adata_fraction = scparadise.scnoah.get_frac(
    path = '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad',
    fraction = 25000,
    stratify = 'cell_type', # Use to get the same ratio as in the full dataset
    random_state = 0
)

In [6]:
# Get raw counts from adata_fraction.raw
adata_fraction = adata_fraction.raw.to_adata()
# Replace variable names with gene names
adata_fraction.var.set_index('feature_name', inplace = True)
adata_fraction.var_names = adata_fraction.var_names.astype('str')
adata_fraction.var_names_make_unique()
# Normalize data
sc.pp.normalize_total(adata_fraction, target_sum = None)
sc.pp.log1p(adata_fraction)

In [7]:
# Select genes for model training (marker genes of cell types)
lst_genes = []
lst_annotations = ['majorclass', 'cell_type'] # annotation levels
for annotation in lst_annotations:
    sc.tl.rank_genes_groups(adata_fraction, 
                            groupby = annotation,
                            method = 't-test_overestim_var', pts = True)
    # Filter marker genes of cell types
    sc.tl.filter_rank_genes_groups(adata_fraction, 
                                   min_fold_change = 1.0, 
                                   min_in_group_fraction = 0.4,
                                   key_added = 'filtered_rank_genes_groups')
    # Create list of genes for model training

    for i in adata_fraction.obs[annotation].unique():
        df = sc.get.rank_genes_groups_df(adata_fraction, group = i, key = 'filtered_rank_genes_groups', pval_cutoff = 0.05)
        df['pts_comparizon'] = df['pct_nz_group']/df['pct_nz_reference']
        lst_genes.extend(df.sort_values(by = 'logfoldchanges', ascending = False).head(20)['names'].tolist())
        lst_genes.extend(df.sort_values(by = 'pts_comparizon', ascending = False).head(20)['names'].tolist())
# Remove duplicates 
lst_genes = np.unique(lst_genes).tolist()
print(f'Number of selected features: {len(lst_genes)}')

Number of selected features: 595


In [8]:
# Subset genes for model training
adata_fraction = adata_fraction[:, lst_genes]

In [9]:
# Alternative way to select genes for model training
# sc.pp.highly_variable_genes(adata_fraction,
#                             n_top_genes = 1000,
#                             subset = True)
# lst_genes = adata_fraction.var_names.tolist()

In [9]:
adata_balanced = scparadise.scnoah.balance(
    adata_fraction, 
    celltype_keys = lst_annotations
)

In [10]:
# Train scadam model using adata_fraction dataset
scparadise.scadam.train(
    adata_balanced,
    path = 'snRNAseq_human_retina', # path to save model
    model_name = 'Human_Retina_scAdam', # folder name with model
    celltype_keys = lst_annotations,
    eval_metric = ['accuracy', 'balanced_accuracy']
)

Device: cuda
Number of features: 595
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20716 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Training scAdam model:   9%|██████▉                                                                      | 18/200 [00:48<08:06,  2.67s/it]

Early stopping triggered! Best score: 0.9924
Training completed!
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam


## Evaluation of model quality

For model evaluation, we use another subset of 25,000 cells generated using a different random state.

In [11]:
# Get test dataset for model quality evaluation
adata_test = scparadise.scnoah.get_frac(
    path = '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad',
    fraction = 25000,
    stratify = 'cell_type',
    random_state = 42
)

In [12]:
# Check common cells between test and training datasets
lst_train = adata_fraction.obs_names.tolist()
lst_test = adata_test.obs_names.tolist()
lst_train.extend(lst_test)
lst_train = np.unique(lst_train)
percent = round((2 * len(lst_test) - len(lst_train))/len(lst_test)*100, 5)
print(f"There are {percent} % common cells ({2 * len(lst_test) - len(lst_train)} cells) between the test and training datasets")

There are 0.836 % common cells (209 cells) between the test and training datasets


Less than 1% of cells are the same between the test dataset and the training dataset. <br>
This number of similar cells can be ignored, and we can proceed with testing the model's quality.

In [13]:
# Apply the same preprocessing steps to the test dataset as used for training
# Get raw counts from adata_fraction.raw
adata_test = adata_test.raw.to_adata()

# Replace variable names with gene names
adata_test.var.set_index('feature_name', inplace = True)
adata_test.var_names = adata_test.var_names.astype('str')
adata_test.var_names_make_unique()

# Normalize data
sc.pp.normalize_total(adata_test, target_sum = None)
sc.pp.log1p(adata_test)
adata_test.raw = adata_test

In [14]:
# Predict cell types using trained model
adata_test = scparadise.scadam.predict(
    adata_test, 
    path_model = 'snRNAseq_human_retina/Human_Retina_scAdam'
)

scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Gene alignment:
  Model features: 595
  Matched features: 595 (100.0%)


Predicting: 100%|████████████████████████████████████████████████████████████████████████████████████████| 98/98 [00:00<00:00, 151.49it/s]


Added cell type column: pred_celltype_l1
Added probabilities column: pred_celltype_l1_probability
Added cell type column: pred_celltype_l2
Added probabilities column: pred_celltype_l2_probability


In [15]:
## Check model quality
df_l1 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype = 'majorclass',
    pred_celltype = 'pred_celltype_l1'
)
df_l1

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
AC,0.9989,0.9978,0.9998,0.9983,0.9988,0.9973,4496
Astrocyte,1.0000,0.982,1.0,0.9909,0.991,0.9802,111
BC,0.9993,0.9998,0.9998,0.9995,0.9998,0.9996,5437
Cone,0.9990,1.0,1.0,0.9995,1.0,1.0,1000
HC,0.9953,0.9984,0.9999,0.9969,0.9991,0.9982,634
MG,0.9994,0.9994,1.0,0.9994,0.9997,0.9993,1744
Microglia,1.0000,0.9744,1.0,0.987,0.9871,0.9719,39
RGC,0.9978,0.999,0.9997,0.9984,0.9994,0.9987,3144
RPE,0.8750,1.0,1.0,0.9333,1.0,1.0,7
Rod,0.9999,0.9995,0.9999,0.9997,0.9997,0.9994,8388


In [16]:
## Check model quality
df_l2 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype = 'cell_type',
    pred_celltype = 'pred_celltype_l2'
)
df_l2

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
GABAergic amacrine cell,0.9944,0.9968,0.9993,0.9956,0.9981,0.9959,2855
H1 horizontal cell,0.9963,0.987,0.9999,0.9916,0.9935,0.9857,540
H2 horizontal cell,0.9192,0.9681,0.9997,0.943,0.9838,0.9647,94
Mueller cell,0.9994,1.0,1.0,0.9997,1.0,1.0,1744
OFF midget ganglion cell,0.8699,0.9708,0.9902,0.9176,0.9805,0.9595,1577
OFF parasol ganglion cell,0.9740,0.9494,0.9999,0.9615,0.9743,0.9445,79
OFFx cell,0.9835,0.9754,0.9999,0.9794,0.9876,0.9729,122
ON midget ganglion cell,0.9652,0.8357,0.9985,0.8958,0.9135,0.8209,1193
ON parasol ganglion cell,1.0000,0.9796,1.0,0.9897,0.9897,0.9776,49
ON-blue cone bipolar cell,1.0000,0.7826,1.0,0.878,0.8847,0.7656,23


The model performs well except for the `retinal ganglion cell`. <br> 

You could try using a different random state to generate another test dataset.

## Iterative warm start training (optional)

You may use another subset of the whole dataset to increase model generalization.

In [18]:
# Do not set the lower bound of the range to 0, which was used for the primary training of the model
for i in range(1, 4):
    # Obtain 25000 cells randomly
    adata_fraction = scparadise.scnoah.get_frac(
        path = '2e910e62-7eaf-4c06-80cb-8918e3eea16e.h5ad',
        fraction = 25000,
        stratify = 'cell_type',
        random_state = i
    )
    
    # Get raw counts from adata_fraction.raw
    adata_fraction = adata_fraction.raw.to_adata()
    
    # Replace variable names with gene names
    adata_fraction.var.set_index('feature_name', inplace=True)
    adata_fraction.var_names = adata_fraction.var_names.astype('str')
    adata_fraction.var_names_make_unique()
    
    # Normalize data
    sc.pp.normalize_total(adata_fraction, target_sum = None)
    sc.pp.log1p(adata_fraction)
    adata_fraction.raw = adata_fraction
    
    # Subset genes for model training
    adata_fraction = adata_fraction[:, lst_genes]
    # Balance dataset
    adata_balanced = scparadise.scnoah.balance(
        adata_fraction, 
        celltype_keys = lst_annotations
    )
    adata_balanced.raw = adata_balanced
    # Warm start requires second training dataset and path to pretrained model 
    scparadise.scadam.warm_start(
        adata_balanced,
        path = 'snRNAseq_human_retina', # path to save model
        path_model = 'snRNAseq_human_retina/Human_Retina_scAdam', # folder name with pretrained model
        model_name = 'Human_Retina_scAdam',
        celltype_keys = lst_annotations,
        eval_metric = ['accuracy', 'balanced_accuracy']
    )

scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Device: cuda
Gene alignment:
 Number of features: 595
 Matched features: 595 (100.0%)
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20716 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Warm-start model fine-tuning:  42%|███████████████████▎                          | 42/100 [02:01<02:47,  2.89s/it]

Early stopping triggered! Best score: 0.9943
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam
scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Device: cuda
Gene alignment:
 Number of features: 595
 Matched features: 595 (100.0%)
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20716 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Warm-start model fine-tuning:  22%|██████████                                    | 22/100 [01:07<03:57,  3.05s/it]

Early stopping triggered! Best score: 0.9932
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam
scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Device: cuda
Gene alignment:
 Number of features: 595
 Matched features: 595 (100.0%)
Label hierarchy: majorclass → cell_type
Annotation levels weights using strategy 'linear_offset':
  majorclass: 10 cell types, 0.4 relative weight
  cell_type: 31 cell types, 0.6 relative weight

Dataset split:
Train dataset contains: 20716 cells, it is 80.0 % of input dataset
Validation dataset contains: 5180 cells, it is 20.0 % of input dataset


Warm-start model fine-tuning:  14%|██████▍                                       | 14/100 [00:34<03:34,  2.50s/it]

Early stopping triggered! Best score: 0.9953
Fitting unknown cells detector


UnknownCellDetector fitted successfully!
Model saved to snRNAseq_human_retina/Human_Retina_scAdam


In [19]:
# Predict cell types using trained model
adata_test = scparadise.scadam.predict(
    adata_test, 
    path_model = 'snRNAseq_human_retina/Human_Retina_scAdam'
)

scAdam model with unknown detector loaded from snRNAseq_human_retina/Human_Retina_scAdam
Gene alignment:
  Model features: 595
  Matched features: 595 (100.0%)


Predicting: 100%|████████████████████████████████████████████████████████████████| 98/98 [00:00<00:00, 133.10it/s]


Added cell type column: pred_celltype_l1
Added probabilities column: pred_celltype_l1_probability
Added cell type column: pred_celltype_l2
Added probabilities column: pred_celltype_l2_probability


In [20]:
## Check model quality
df_warm_start_l1 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype='majorclass',
    pred_celltype='pred_celltype_l1'
)
df_warm_start_l1

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
AC,0.9987,0.9998,0.9997,0.9992,0.9997,0.9995,4496
Astrocyte,0.9909,0.982,1.0,0.9864,0.9909,0.9802,111
BC,0.9998,0.9996,0.9999,0.9997,0.9998,0.9995,5437
Cone,1.0000,0.999,1.0,0.9995,0.9995,0.9989,1000
HC,1.0000,1.0,1.0,1.0,1.0,1.0,634
MG,1.0000,1.0,1.0,1.0,1.0,1.0,1744
Microglia,1.0000,0.9744,1.0,0.987,0.9871,0.9719,39
RGC,0.9990,0.9994,0.9999,0.9992,0.9996,0.9992,3144
RPE,1.0000,1.0,1.0,1.0,1.0,1.0,7
Rod,0.9998,0.9995,0.9999,0.9996,0.9997,0.9994,8388


In [21]:
## Check model quality
df_warm_start_l2 = scparadise.scnoah.report_classif_full(
    adata_test,
    celltype='cell_type',
    pred_celltype='pred_celltype_l2'
)
df_warm_start_l2

,precision,recall/sensitivity,specificity,f1-score,geometric mean,index balanced accuracy,number of cells
GABAergic amacrine cell,0.9965,0.9954,0.9995,0.996,0.9975,0.9946,2855
H1 horizontal cell,0.9981,0.9944,1.0,0.9963,0.9972,0.9939,540
H2 horizontal cell,0.9688,0.9894,0.9999,0.9789,0.9946,0.9882,94
Mueller cell,1.0000,1.0,1.0,1.0,1.0,1.0,1744
OFF midget ganglion cell,0.9290,0.9461,0.9951,0.9375,0.9703,0.9369,1577
OFF parasol ganglion cell,0.9630,0.9873,0.9999,0.975,0.9936,0.986,79
OFFx cell,0.9918,0.9918,1.0,0.9918,0.9959,0.991,122
ON midget ganglion cell,0.9587,0.9346,0.998,0.9465,0.9658,0.9268,1193
ON parasol ganglion cell,1.0000,0.9796,1.0,0.9897,0.9897,0.9776,49
ON-blue cone bipolar cell,1.0000,0.8696,1.0,0.9302,0.9325,0.8582,23


In [22]:
pd.set_option('display.max_rows', 100)
df_l2.compare(df_warm_start_l2, keep_equal=True, align_axis = 0, result_names=('default', 'warm start'))

precision recall/sensitivity  \
GABAergic amacrine cell          default        0.9934             0.9961   
                                 warm start     0.9965             0.9954   
H1 horizontal cell               default        0.9926             0.9944   
                                 warm start     0.9981             0.9944   
H2 horizontal cell               default        0.9677             0.9574   
                                 warm start     0.9688             0.9894   
Mueller cell                     default        0.9994             0.9994   
                                 warm start     1.0000                1.0   
OFF midget ganglion cell         default        0.9079             0.9562   
                                 warm start     0.9290             0.9461   
OFF parasol ganglion cell        default        0.9487             0.9367   
                                 warm start     0.9630             0.9873   
OFFx cell                        default        0.9835             0.9754   
                                 warm start     0.9918             0.9918   
ON midget ganglion cell          default        0.9521             0.9162   
                                 warm start     0.9587             0.9346   
ON parasol ganglion cell         default        0.9796             0.9796   
                                 warm start     1.0000             0.9796   
S cone cell                      default        0.9844             0.9403   
                                 warm start     0.9571                1.0   
amacrine cell                    default        0.9863             0.9622   
                                 warm start     0.9822               0.98   
astrocyte                        default        1.0000              0.982   
                                 warm start     0.9909              0.982   
diffuse bipolar 1 cell           default        0.9975             0.9874   
                                 warm start     0.9949             0.9924   
diffuse bipolar 2 cell           default        0.9926             0.9963   
                                 warm start     0.9925             0.9925   
diffuse bipolar 3a cell          default        1.0000             0.9826   
                                 warm start     0.9942             0.9942   
diffuse bipolar 3b cell          default        0.9964             0.9928   
                                 warm start     0.9858             0.9964   
diffuse bipolar 4 cell           default        0.9923                1.0   
                                 warm start     0.9948             0.9948   
diffuse bipolar 6 cell           default        0.9931             0.9863   
                                 warm start     0.9863             0.9863   
flat midget bipolar cell         default        0.9956             0.9974   
                                 warm start     0.9982             0.9956   
giant bipolar cell               default        0.9802              0.995   
                                 warm start     0.9851                1.0   
glycinergic amacrine cell        default        0.9790             0.9922   
                                 warm start     0.9836             0.9903   
invaginating midget bipolar cell default        0.9964             0.9964   
                                 warm start     0.9964             0.9988   
microglial cell                  default        1.0000             0.9487   
                                 warm start     1.0000             0.9744   
retinal bipolar neuron           default        0.9951             0.9951   
                                 warm start     0.9976             0.9927   
retinal cone cell                default        0.9947             0.9989   
                                 warm start     1.0000             0.9957   
retinal ganglion cell            default        0.5749             0.4837   
                                 warm sta

Iterative warm start training led to an increase in all model quality metrics (rows - macro average, weighted average, accuracy, and balanced accuracy). <br>
Additionally, the model's sensitivity increased by 10.17% and precision by 0.81% for the `retinal ganglion cell`.

In [23]:
import session_info
session_info.show()